### DISCUSSION TITLE
'framebuf' display simulation

### DISCUSSION THREAD

**Discussion by fma38** [micropython/micropython#11715] [Unix & Windows]
Hi!

Is there a graphical library available for Micropython Unix port ? Like tkinter/PIL...

I would like to simulate a framebuf display (like SSD1306), so I can develop and test a GUI lib without using a real microcontroller/screen/buttons...

Thanks, 

Frédéric

**Comment by 2dof** [comment]
Try pygame as main "display engine" for create simulator, or  try [luma.emulator](https://github.com/rm-hull/luma.examples/tree/master)


edit: [luma.oled](https://github.com/rm-hull/luma.oled)

**Comment by fma38** [comment]
Thanks for your answer! luma.emulator looks nice...
But how do I install pygame or luma in the micropython unix port?

**Reply by 2dof** [reply]
sorry, my mistake, to I thought that You are looking for some emulator for python (not micro python) Rasspbbery pi platform, or even in Macos/win.

 I used  luma for oled 0.91 display  emulation for design menu system for esp32. ; python pynput.keyboard for  keyboard (buttons) interaction ( or just call callback function for keys - even better since all can be done from program).

**Comment by fma38** [comment]
Well, my goal is to work on a micropython menu/gui lib which uses displays like ssd1306 and similar, which are based on framebuf. That's why I wanted to use micropython unix port.

But I may be able to use regular Python, by emulating the entire framebuf module...

**Reply by peterhinch** [reply]
There is no support for this way of working. You'll have to use real hardware or write your own support.

I guess you could write an emulator using (e.g.) pygame and CPython, but then you'd need to write a `FrameBuffer` class for CPython (the MP class is written in C). A nontrivial task. I'm not convinced an emulator would save much time in practice. With the right tools (`mpremote`, `rshell`) development with real hardware can be quite quick.

**Reply by 2dof** [reply]
luma has a lot of SSD family drivcers simulators, in  my case i used to emulate ssd1306 oleds  0.91/0.96 inch do design  menu system for esp32 micropython,  but all in python (luma was deeded to all fancy 'icons' drawnings). 

 Menu/GUI system  can be 'split' for :  display rendering and for interaction with peripherals ( buttons, etc)  -> all this part need to be implemented  as some  display rendering functions and some callback functions for interaction ( callback "connected" to button action).  In simplest way we do not even need physical button, just call a callback function responsible for assigned button action.  
Most of work is just some classes implementing and  can be done in python  (just need to know difference betwen  micropython-vs-python) for faster R&D .   


edit. I forgot to mention than luma will not work on windows system ( no ioctl).  
       other method for modelling  oled display is using [python PIL lib](https://www.bluetin.io/displays/create-oled-lcd-pillow-graphics-fonts/)

**Comment by fma38** [comment]
The problem is to have acces to the hardware: I would like to be able to develop from anywhere, when I have some free time (in train, for example)...
Is it complicated to recompile a micropython module written in C for regular Python?

**Reply by peterhinch** [reply]
The whole thing is a substantial task. Rewriting C code in Python is straightforward but time consuming. Then you have to add the pygame stuff.

I would compare the effort to buying a display which supports a Pico plugged in the back and mounting it in a box. Plug the USB lead into your laptop and develop anywhere. That is my approach, but the choice is yours.

**Reply by fma38** [reply]
Yes, I'm not sure to want to dig into C that much!
As you say, it is more fun to build a real hardware stuff ;o)
Thanks,

**Reply by peterhinch** [reply]
The file [modframebuf.c](https://github.com/micropython/micropython/blob/master/extmod/modframebuf.c) is quite substantial. While an emulator would be an interesting project it would take quite a lot of writing and testing. If the emulation wasn't spot-on it would be more trouble than it's worth.

My view of MP development is to use real hardware with simple command-line tools.

**Comment by rkompass** [comment]
I'm assuming you use the Linux OS. The framebuffer memory is a chunk of RAM which is pixel-wise arranged.
You may query its address from within MP.
I naively think it wouldn't be too difficult to write a C-program using the Cairo library, for example, that reads this chunk of RAM, re-codes the pixel values, if necessary, and displays them in a window. Including permanent refreshes, of course.
But I'm not sure. These would be different processes. Can they access the memory of each other?

**Reply by fma38** [reply]
I don't even think there is any pixel to rearrange, even using tkinter canvas + third-party PhotoImageTk or so (I read quickly some examples).
But I don't think it is possible to exchange memory between MP and external program (either in C or CPython)...

**Reply by rkompass** [reply]
It seems possible, theoretically.
[https://gist.github.com/FergusInLondon/fec6aebabc3c9e61e284983618f40730](https://gist.github.com/FergusInLondon/fec6aebabc3c9e61e284983618f40730).

**Reply by fma38** [reply]
Interesting!

**Reply by 2dof** [reply]
I have googled some [online simulator](https://wokwi.com/projects/305568836183130690), unfortunately internet connection deeded.

**Reply by russhughes** [reply]
I played around with the Unix port, and I was able to create a USER_C_MODULE that uses SDL2 to display the contents of a `framebuf.FrameBuffer` on a monitor.  I'll work on it more this weekend; if it looks useful, I'll post it to Git Hub.

**Reply by fma38** [reply]
That would be great!

**Reply by russhughes** [reply]
I'm planning on cleaning it up this weekend: https://www.youtube.com/shorts/2p4RyBjm_-E

**Reply by fma38** [reply]
That's brilliant! This will be a very powerful tool :o)

**Reply by rkompass** [reply]
This looks very very nice.
@russhughes big thanks!  Did you perhaps plan to have an option to zoom the pixels out (like 1 pixel --> 3x3 pixels on Monitor)?
Otherwise I might have difficulties to see the details on my screen.

**Reply by russhughes** [reply]
@rkompass Yes, I plan to support X & Y scaling since displays have different pixel aspect ratios.

**Reply by russhughes** [reply]
The driver is my first SLD2 project, so there is room for improvement. It works for me(tm). I tested it running under Almalinux 9 and on Ubuntu 20.04.5 LTS and 22.04 LTS using WSL2 on Windows 11. In the WSL2 environments, the demos core dump when you close the display window. I don't know if this is an issue with my code or if the WSL2 environment causes this.  

The driver is available at [SDL2 Display driver for the MicroPython Unix port.](https://github.com/russhughes/micropython_sdl2).

**Reply by fma38** [reply]
It works great, thanks!
I need to dig into your examples, but I'm pretty sure it will be very easy to adapt them to my use case.
Again, great job!

**Reply by mattytrentini** [reply]
That's awesome work @russhughes, thanks for your efforts! Could definitely have used this in the past.

Not to take anything, at all!, away from your excellent work, but it sparked a few (some old) thoughts and I thought I'd share: 
* It would be interesting to see if it were possible to make this a [native module](https://docs.micropython.org/en/latest/develop/natmod.html) to reduce the build friction and ease deployment.
* I wonder if this could also be implemented with [ffi](https://github.com/micropython/micropython-lib/tree/master/unix-ffi). Could possibly be written entirely in Python. Even easier to deploy.
* I've wanted a similar solution, but to render the output to an HTML canvas, with data transferred over a web socket. The benefit would be that it could also be used easily from within a container.

**Reply by fma38** [reply]
+1 for native module!

**Reply by russhughes** [reply]
@mattytrentini @fma38 I'll look into it.

**Reply by russhughes** [reply]
@mattytrentini @fma38 I don't see a way to link an external library like SDL2 with a native .mpy module without modifying MicroPython. I'm temped to try adding dlopen, dlsyn, dlclose, and dlerror to the mp_fun_table to see if that is a way around natmod linking limitations.

**Comment by massimosala** [comment]
Hi

I was intrigued by this problem and it is a feature useful to me too.
So, between Monday and yesterday, I develop a virtual driver.

At the moment I use to replace the real ssd1306 driver.
It supports mono and color displays, of any size.
To not overload the micro memory, the framebuffer is created on the web page :-)

So it works only with web applications, I mean micropython programs with a web server.

It consists of two files:
- the micropython module, about 5 Kb (1.3 Kb compiled to MPY);
- the javascript module (rendering the display on the web page), about 10 Kb.


I am thinking to publish it as open source, but I have to improve some features.
Any of you want to be a beta tester?
Write to massimo.sala.71@gmail.com
